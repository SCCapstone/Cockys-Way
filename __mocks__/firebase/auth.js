export const getAuth = jest.fn(() => ({
    currentUser: {
      uid: 'test-user',
      email: 'test@example.com',
    },
  }));
  