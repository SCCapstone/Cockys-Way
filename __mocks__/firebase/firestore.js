export const getFirestore = jest.fn();
export const collection = jest.fn();
export const doc = jest.fn();

export const getDocs = jest.fn((ref) => {
  if (ref._path?.segments?.includes("assignments")) {
    return Promise.resolve({
      forEach: (callback) => {
        callback({
          data: () => ({
            title: "Midterm Essay",
            courseCode: "CS101",
            dueDate: new Date("2024-09-20"),
          }),
        });
      },
    });
  }

  return Promise.resolve({
    docs: [
      {
        id: "course1",
        data: () => ({
          name: "Test Course",
          code: "CS101",
          section: "001",
        }),
      },
    ],
  });
});

export const getDoc = jest.fn(() =>
  Promise.resolve({
    exists: () => false,
    data: () => ({}),
  })
);

export const updateDoc = jest.fn();
export const addDoc = jest.fn(() => Promise.resolve({ id: "mock-id" }));
export const deleteDoc = jest.fn();
