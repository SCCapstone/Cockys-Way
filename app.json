{
  "expo": {
    "name": "Cockys-Way",
    "slug": "Cockys-Way",
    "version": "1.0.0",
    "orientation": "portrait",
    "icon": "./assets/icon.png",
    "userInterfaceStyle": "light",
    "splash": {
      "image": "./assets/splash.png",
      "resizeMode": "contain",
      "backgroundColor": "#ffffff"
    },
    "ios": {
      "supportsTablet": true
    },
    "android": {
      "adaptiveIcon": {
        "foregroundImage": "./assets/adaptive-icon.png",
        "backgroundColor": "#ffffff"
      },
      "package": "com.ryanmalone.CockysWay"
    },
    "web": {
      "favicon": "./assets/favicon.png"
    },
    "extra": {
      "eas": {
        "projectId": "35142afc-1c33-4347-85b2-0c264a39eab2"
      }
    }
  }
}
