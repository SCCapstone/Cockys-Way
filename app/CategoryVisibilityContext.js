// Using this for Pin Filters for the Map

import React, { createContext, useState, useEffect } from "react";

export const CategoryVisibilityContext = createContext();

export const CategoryVisibilityProvider = ({ children, markers }) => {
  const [categoryVisibility, setCategoryVisibility] = useState({});
  const [isInitialized, setIsInitialized] = useState(false); // changing it so default shown categories are from here

  useEffect(() => {
    const initializeVisibility = () => {
      //  console.log("Initializing category visibility..."); // Debugging
      const initialVisibility = {};

      // make just colleges & schools shown by default
      const collegesAndSchoolsCatIds = [
        24560, 24912, 24903, 24904, 24905, 24914, 24907, 24902, 24908, 24906,
        24909, 24910, 24911, 24913, 24901,
      ];

      collegesAndSchoolsCatIds.forEach((id) => {
        initialVisibility[id] = true; // Colleges & Schools visible
      });

      // set others to hidden
      if (markers && markers.length > 0) {
        markers.forEach((marker) => {
          if (!collegesAndSchoolsCatIds.includes(marker.catId)) {
            initialVisibility[marker.catId] = false; // Hidden
          }
        });
      } //else {
      //  console.warn("Markers array is empty or undefined."); // Debugging
      //}

      setCategoryVisibility(initialVisibility);
      setIsInitialized(true); // Initialization should be DONE.
      // console.log("Category visibility initialized:", initialVisibility); // debug
    };
    //console.log("Initialized 2: ", isInitialized);    // Debugging

    if (!isInitialized) {
      initializeVisibility();
    }
  }, [markers, isInitialized]);

  // Log `isInitialized` whenever it changes
  useEffect(() => {
    // console.log("isInitialized updated:", isInitialized);
  }, [isInitialized]);

  return (
    <CategoryVisibilityContext.Provider
      value={{ categoryVisibility, setCategoryVisibility, isInitialized }}
    >
      {children}
    </CategoryVisibilityContext.Provider>
  );
};

export default CategoryVisibilityProvider;
