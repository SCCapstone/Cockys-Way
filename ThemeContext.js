import React, { createContext, useEffect, useState } from "react";
import { getAuth, onAuthStateChanged } from "firebase/auth";
import { doc, getDoc } from "firebase/firestore";
import { FIRESTORE_DB } from "./FirebaseConfig";
import * as SplashScreen from "expo-splash-screen";

export const ThemeContext = createContext();

const customLightTheme = {
  dark: false,
  colors: {
    primary: "#73000A", // garnet
    garnetWhite: "#73000A",
    background: "#FFFFFF", // white
    card: "#F2F2F2", // light gray
    text: "#000000", // black text
    border: "#E0E0E0", // light border
    notification: "#73000A", // match primary
    alwaysWhite: "#FFFFFF", // white
    placeholder: "#A9A9A9", // light gray placeholder
    lightGarnet: "#D5B4BA", // light garnet
    disabledDay: "#B6BCC2" // very light gray
  },
};

const customDarkTheme = {
  dark: true,
  colors: {
    primary: "#73000A", // dark version still using garnet
    garnetWhite: "#FFFFFF",
    background: "#757474", // deep dark background
    card: "#2C2C2C", // card/drawer background
    text: "#E0E0E0", // white text
    border: "#5a5a5a", // darker border
    notification: "#FF8A80", // lighter red
    alwaysWhite: "#FFFFFF", // white
    placeholder: "#A9A9A9", // light
    lightGarnet: "#9C6B70", // light garnet
    disabledDay: "#75797D" // darker gray
  },
};

export const ThemeProvider = ({ children }) => {
  const [isDarkTheme, setIsDarkTheme] = useState(false);
  const [isLoaded, setIsLoaded] = useState(false);

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(getAuth(), async (user) => {
      if (!user) {
        setIsLoaded(true);
        await SplashScreen.hideAsync();
        return;
      }

      try {
        const docRef = doc(FIRESTORE_DB, "settings", user.uid);
        const docSnap = await getDoc(docRef);

        if (docSnap.exists()) {
          const userSettings = docSnap.data().settings;
          if (userSettings?.theme === "dark") {
            setIsDarkTheme(true);
          }
        }
      } catch (error) {
        console.error("Error loading theme from Firestore:", error);
      }

      setIsLoaded(true);
      await SplashScreen.hideAsync();
    });

    return () => unsubscribe();
  }, []);

  const theme = isDarkTheme ? customDarkTheme : customLightTheme;

  return (
    <ThemeContext.Provider
      value={{ isDarkTheme, setIsDarkTheme, theme, isLoaded }}
    >
      {isLoaded ? children : null}
    </ThemeContext.Provider>
  );
};
