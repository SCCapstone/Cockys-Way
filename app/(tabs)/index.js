import React, { useEffect, useState, useRef } from "react";
import MapView, { Marker, PROVIDER_GOOGLE } from "react-native-maps";
import {
  StyleSheet,
  SafeAreaView,
  Alert,
  View,
  Text,
  TouchableOpacity,
  ScrollView,
} from "react-native";
import { collection, getDocs } from "firebase/firestore";
import { FIRESTORE_DB } from "../../FirebaseConfig";
import { useRouter } from "expo-router";
import { SearchBar } from "react-native-elements";
import * as SplashScreen from "expo-splash-screen";
import * as Location from "expo-location";
import MapViewDirections from "react-native-maps-directions";
import { GOOGLE_API_KEY } from "@env";
import styles from "../../homestyles";

// Prevent the splash screen from auto-hiding
SplashScreen.preventAutoHideAsync();

export default function HomeScreen() {
  const router = useRouter();
  const [markers, setMarkers] = useState([]);
  const [filteredMarkers, setFilteredMarkers] = useState([]);
  const [search, setSearch] = useState("");
  const [userLocation, setUserLocation] = useState(null);
  const [selectedDestination, setSelectedDestination] = useState(null);
  const [travelMode, setTravelMode] = useState("DRIVING");
  const [startLocation, setStartLocation] = useState(null);
  const [routeDetails, setRouteDetails] = useState(null);
  const [showTravelModeButtons, setShowTravelModeButtons] = useState(false);
  const [showRouteDetails, setShowRouteDetails] = useState(false);
  const [showTraffic, setShowTraffic] = useState(false);
  const [routeSteps, setRouteSteps] = useState([]);
  const mapRef = useRef(null);
  const [isLoading, setIsLoading] = useState(true); // testing to see if loading works?

  const INITIAL_REGION = {
    latitude: 34.00039991787572,
    longitude: -81.03594096158815,
    latitudeDelta: 0.05,
    longitudeDelta: 0.05,
  };

  const onMarkerSelected = (marker) => {
    Alert.alert(marker.title || "Marker Selected");

    setSelectedDestination({
      latitude: marker.latitude,
      longitude: marker.longitude,
    });

    setShowTravelModeButtons(true);
    setShowRouteDetails(true);

    // Zoom in on marker region
    if (mapRef.current) {
      mapRef.current.animateToRegion(
        {
          latitude: marker.latitude,
          longitude: marker.longitude,
          latitudeDelta: 0.01,
          longitudeDelta: 0.01,
        },
        2500 // 2500 is duration of zoom in ms
      );
    }
  };

  // Fetch markers from Firebase
  useEffect(() => {
    const fetchMarkers = async () => {
      try {
        //const query = await getDocs(collection(FIRESTORE_DB, "markers"));   // OG
        const query = await getDocs(collection(FIRESTORE_DB, "locTest")); // Changed to use what Chloe brought in
        const db_data = query.docs.map((doc) => ({
          id: doc.id,
          ...doc.data(),
        }));

        setMarkers(db_data);
        setFilteredMarkers(db_data);
        setIsLoading(false);
      } catch (err) {
        Alert.alert("Error fetching data");
        setIsLoading(false);
      }
    };

    fetchMarkers();
  }, []);

  // Update filtered markers based on search input
  useEffect(() => {
    if (search === "") {
      setFilteredMarkers(markers);
    } else {
      const filtered = markers.filter((marker) =>
        marker.title.toLowerCase().includes(search.toLowerCase())
      );
      setFilteredMarkers(filtered);
    }
  }, [search, markers]);

  // Request location permissions and set startLocation
  // merged both perms and set user location into one -Isaac
  useEffect(() => {
    const getLocation = async () => {
      const { status } = await Location.requestForegroundPermissionsAsync();
      if (status === "granted") {
        const location = await Location.getCurrentPositionAsync({});
        setUserLocation(location.coords);
        setStartLocation(location.coords);
      } else {
        Alert.alert("Permission denied, please grant location access");
      }
    };

    getLocation();
  }, []);

  // Hide splash screen
  useEffect(() => {
    async function prepare() {
      try {
        await SplashScreen.hideAsync();
      } catch (e) {
        console.warn(e);
      }
    }

    prepare();
  }, []);

  // Handle stopping directions
  const handleStopDirections = () => {
    setSelectedDestination(null);
    setShowTravelModeButtons(false);
    setShowRouteDetails(false);
    setRouteDetails(null);
  };

  // Change the start location to selected marker
  const handleChangeStartLocation = () => {
    if (selectedDestination) {
      setStartLocation({
        latitude: selectedDestination.latitude,
        longitude: selectedDestination.longitude,
      });
    }
  };

  // Reset start location to user's current location
  const handleResetStartLocation = () => {
    if (userLocation) {
      setStartLocation(userLocation);
    }
  };

  if (isLoading) {
    return (
      <View style={styles.loadingContainer}>
        <Text>Loading...</Text>
      </View>
    );
  }

  return (
    <SafeAreaView style={styles.container}>
      <SearchBar
        placeholder="Search Here..."
        onChangeText={(text) => setSearch(text)}
        value={search}
        containerStyle={styles.searchContainer}
        inputContainerStyle={styles.searchInputContainer}
      />
      <View style={styles.buttonContainer}>
        <TouchableOpacity
          style={styles.filterButton}
          onPress={() => router.push("/PinFilterMain")}
        >
          <View style={styles.accentBox}>
            <Text style={styles.filterButtonText}>Filter Pins</Text>
          </View>
        </TouchableOpacity>

        {/* Button to show/hide traffic */}
        <TouchableOpacity
          style={styles.trafficButton}
          onPress={() => setShowTraffic(!showTraffic)}
        >
          <Text style={styles.trafficButtonText}>
            {showTraffic ? "Hide Traffic" : "Show Traffic"}
          </Text>
        </TouchableOpacity>
      </View>

      {/* Map */}
      <MapView
        ref={mapRef}
        style={styles.map}
        apiKey={GOOGLE_API_KEY}
        provider={PROVIDER_GOOGLE}
        initialRegion={INITIAL_REGION}
        showsUserLocation={true}
        followsUserLocation={true}
        showsTraffic={showTraffic}
      >
        {filteredMarkers.map((marker) => (
          <Marker
            key={marker.id}
            coordinate={{
              latitude: marker.latitude,
              longitude: marker.longitude,
            }}
            title={marker.title}
            description={marker.description}
            pinColor={marker.color ? marker.color : "red"}
            onPress={() => onMarkerSelected(marker)}
          />
        ))}

        {/* Directions */}
        {startLocation && selectedDestination && (
          <MapViewDirections
            origin={startLocation}
            destination={{
              latitude: selectedDestination.latitude,
              longitude: selectedDestination.longitude,
            }}
            apikey={GOOGLE_API_KEY}
            strokeWidth={4}
            strokeColor="#73000a"
            mode={travelMode}
            onReady={(result) => {
              setRouteDetails({
                distance: result.distance,
                duration: result.duration,
              });
              setRouteSteps(result.legs[0].steps || []);
            }}
            onError={(error) => Alert.alert("Error getting directions", error)}
          />
        )}
      </MapView>

      {/* Travel Mode Buttons (Overlay) */}
      {showTravelModeButtons && (
        <View style={styles.travelModeOverlay}>
          <TouchableOpacity
            style={[
              styles.travelModeButton,
              travelMode === "DRIVING" && styles.activeTravelMode,
            ]}
            onPress={() => setTravelMode("DRIVING")}
          >
            <Text style={styles.travelModeText}>Driving</Text>
          </TouchableOpacity>
          <TouchableOpacity
            style={[
              styles.travelModeButton,
              travelMode === "WALKING" && styles.activeTravelMode,
            ]}
            onPress={() => setTravelMode("WALKING")}
          >
            <Text style={styles.travelModeText}>Walking</Text>
          </TouchableOpacity>
          <TouchableOpacity
            style={[
              styles.travelModeButton,
              travelMode === "BICYCLING" && styles.activeTravelMode,
            ]}
            onPress={() => setTravelMode("BICYCLING")}
          >
            <Text style={styles.travelModeText}>Bicycling</Text>
          </TouchableOpacity>
        </View>
      )}

      {/* Route Details and Stop Button */}
      {showRouteDetails && routeDetails && (
        <View style={styles.routeDetailsContainer}>
          {/* Total Distance and Duration */}
          <Text style={styles.routeDetailsText}>
            Distance: {routeDetails.distance.toFixed(2)} miles
          </Text>
          <Text style={styles.routeDetailsText}>
            Duration: {Math.ceil(routeDetails.duration)} minutes
          </Text>

          {/* Step-by-Step Instructions */}
          {routeSteps && routeSteps.length > 0 ? (
            <ScrollView style={{ maxHeight: 150 }}>
              {routeSteps.map((step, index) => (
                <View key={index} style={{ marginBottom: 10 }}>
                  <Text style={styles.routeDetailsText}>
                    {step.html_instructions.replace(/<[^>]+>/g, "")}
                  </Text>
                  <Text style={styles.routeDetailsText}>
                    Distance: {step.distance.text}, Duration:{" "}
                    {step.duration.text}
                  </Text>
                </View>
              ))}
            </ScrollView>
          ) : (
            <Text style={styles.routeDetailsText}>No steps available.</Text>
          )}

          {/* Buttons */}
          <TouchableOpacity
            style={styles.changeStartButton}
            onPress={handleChangeStartLocation}
          >
            <Text style={styles.changeStartButtonText}>Set Start Here</Text>
          </TouchableOpacity>
          <TouchableOpacity
            style={styles.changeStartButton}
            onPress={handleResetStartLocation}
          >
            <Text style={styles.changeStartButtonText}>
              Reset to My Location
            </Text>
          </TouchableOpacity>
          <TouchableOpacity
            style={styles.stopButton}
            onPress={handleStopDirections}
          >
            <Text style={styles.stopButtonText}>Stop Directions</Text>
          </TouchableOpacity>
        </View>
      )}
    </SafeAreaView>
  );
}
