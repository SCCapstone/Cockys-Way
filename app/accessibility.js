import React, { useEffect } from "react";
import { View, Text, StyleSheet, ScrollView } from "react-native";
import { FIREBASE_AUTH, FIREBASE_DB } from "../FirebaseConfig";
import { updateProfile, getAuth } from "firebase/auth";
import { doc, setDoc } from "firebase/firestore";
import { router } from "expo-router";

export default function AccessibilityScreen() {
  return (
    <ScrollView style={styles.container}>
      <Text style={styles.header}>Accessibility</Text>
      <View style={styles.content}>
        <Text style={styles.text}>
          Here you can manage your accessibility settings.
        </Text>
        {/* Add more content as needed */}
      </View>
    </ScrollView>
  );
}

const updateAccessibilitySettings = async (newSettings) => {
  const user = getAuth().currentUser;
  if (user) {
    try {
      const userDoc = doc(FIREBASE_DB, "users", user.uid);
      await setDoc(userDoc, newSettings, { merge: true });
      console.log("Accessibility settings updated successfully");
    } catch (error) {
      console.error("Error updating accessibility settings:", error);
    }
  } else {
    console.error("No UID provided");
  }
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
    backgroundColor: "#F3F3F3",
  },
  header: {
    fontSize: 30,
    fontWeight: "bold",
    marginBottom: 20,
    color: "#73000A",
  },
  content: {
    backgroundColor: "#FFFFFF",
    padding: 20,
    borderRadius: 5,
  },
  text: {
    fontSize: 18,
    color: "#000000",
  },
});

export { updateAccessibilitySettings };
