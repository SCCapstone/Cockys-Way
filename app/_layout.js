import { Stack, Slot } from "expo-router";
import { useContext } from "react";
import { ThemeProvider, ThemeContext } from "../ThemeContext";
import { ActivityIndicator, View } from "react-native";
import { DefaultTheme } from "@react-navigation/native";
import { CategoryVisibilityProvider } from "./CategoryVisibilityContext";
import "react-native-get-random-values";
// added above line to connect Map with Pin Filter page for Visibility use

function AppLayout() {
  const { isLoaded, theme } = useContext(ThemeContext);

  if (!isLoaded) {
    return (
      <View style={{ flex: 1, justifyContent: "center", alignItems: "center" }}>
        <ActivityIndicator size="large" />
      </View>
    );
  }

  const screenOptions = {
    headerStyle: {
      backgroundColor: theme.colors.card,
    },
    headerTintColor: theme.colors.text,
  };

  return (
    <Stack screenOptions={screenOptions}>
      <Stack.Screen name="(tabs)" options={{ headerShown: false }} />
      <Stack.Screen
        name="login"
        options={{
          title: "Login",
          headerBackTitle: "Back",
          headerTintColor: theme.colors.text,
        }}
      />
      <Stack.Screen
        name="register"
        options={{
          title: "Register",
          headerBackTitleVisible: false,
          headerTintColor: theme.colors.text,
        }}
      />
      <Stack.Screen
        name="professorInfo"
        options={{
          title: "",
          headerTintColor: theme.colors.text,
          headerBackTitle: "Back",
        }}
      />
      <Stack.Screen
        name="addClassForm"
        options={{
          title: "Search for Classes",
          headerTintColor: theme.colors.text,
          headerBackTitle: "Back",
        }}
      />
      <Stack.Screen
        name="forgotPassword"
        options={{
          title: "Forgot Password",
          headerBackTitleVisible: false,
          headerTintColor: theme.colors.text,
        }}
      />
      <Stack.Screen
        name="addClassSearchResults"
        options={{
          title: "Search Results",
          headerBackTitleVisible: false,
          headerTintColor: theme.colors.text,
        }}
      />
      <Stack.Screen
        name="PrivacySecurity"
        options={{
          title: "Privacy and Security",
          headerBackTitle: "Back",
          headerTintColor: theme.colors.text,
        }}
      />
      <Stack.Screen
        name="favLocations"
        options={{
          title: "Favorite Locations",
          headerTintColor: theme.colors.text,
          headerBackTitle: "Back",
        }}
      />
      <Stack.Screen
        name="MyAccount"
        options={{
          title: "My Account",
          headerTintColor: theme.colors.text,
          headerBackTitle: "Back",
        }}
      />

      <Stack.Screen
        name="PinFilterMain"
        options={{
          title: "Filter Pins",
          headerBackTitle: "Back",
          headerTintColor: theme.colors.text,
        }}
      />
      <Stack.Screen
        name="routeHistory"
        options={{
          title: "Route History",
          headerTintColor: theme.colors.text,
          headerBackTitle: "Back",
        }}
      />
      <Stack.Screen
        name="courseInfo"
        options={{
          title: "Course Info",
          headerTintColor: theme.colors.text,
        }}
      />
      <Stack.Screen
        name="SignOut"
        options={{
          title: "Sign Out",
          headerBackTitle: "Back",
          headerTintColor: theme.colors.text,
        }}
      />
    </Stack>
  );
}

export default function Layout() {
  return (
    <CategoryVisibilityProvider>
      <ThemeProvider>
        <AppLayout />
      </ThemeProvider>
    </CategoryVisibilityProvider>
  );
}

export { theme } from "../ThemeContext";
