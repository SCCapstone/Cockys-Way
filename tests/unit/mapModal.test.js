import React from "react";
import { render, screen, waitFor, fireEvent  } from "@testing-library/react-native";
import HomeScreen from "../../app/(tabs)";
import { ThemeContext } from "../../ThemeContext";
import { CategoryVisibilityContext } from "../../app/CategoryVisibilityContext"; // need this bc otherwise its annoying
import { flatten } from "react-native/Libraries/StyleSheet/StyleSheet";
import { act } from "react-test-renderer";
import { useLocalSearchParams, useRouter } from "expo-router";
import * as Location from "expo-location";
import * as firestore from "firebase/firestore";


/*
        Keeping comments in case working on other tests breaks something
        so I can go back to the start if necessary -Chloe

        run with
        npm --trace-deprecation test -- mapModal
*/

//          MOCKS

// Mock Theme
const mockTheme = {
    colors: {
        text: "#000000",
        garnetWhite: "#FFFFFF",
    },
}; // End theme

// Mock category Visibility
const mockCategoryVisibility = {
    categoryVisibility: {
        9492: true,
        23396: true,
    },
    isInitialized: true,
}; // end category visibility

// Making sure CategoryVisibilityContext is mocked correctly
jest.mock("../../app/CategoryVisibilityContext", () => {
    const React = require("react");
    const actual = jest.requireActual("../../app/CategoryVisibilityContext");
    return {
        ...actual,
        CategoryVisibilityContext: React.createContext({
        categoryVisibility: {
            9492: true,
            23396: true,
        },
        isInitialized: true,
        }),
        useCategoryVisibility: () => ({
        categoryVisibility: {
            9492: true,
            23396: true,
        },
        isInitialized: true,
        }),
    };
}); // end CategoryVisibilityContext


// rendering home screen to make life easier
const renderHomeScreen = (props = {}) => {
    return render(
      <ThemeContext.Provider value={{ theme: mockTheme }}>
        <CategoryVisibilityContext.Provider value={mockCategoryVisibility}>
            <HomeScreen {...props} />
        </CategoryVisibilityContext.Provider>
      </ThemeContext.Provider>
    );
}; // end renderHomeScreen

//              MOCKS

jest.mock("expo-router", () => ({
    useRouter: jest.fn(),
    useLocalSearchParams: jest.fn(),
})); // end mock expo-router

jest.mock("../../FirebaseConfig", () => {
    return {
        FIREBASE_AUTH: {
        currentUser: { uid: "test-user" },
        },
        FIRESTORE_DB: {},
    };
}); // end mock FirebaseConfig

jest.mock("expo-location", () => ({
    requestForegroundPermissionsAsync: jest.fn(() =>
        Promise.resolve({ status: "granted" })
    ),
    getCurrentPositionAsync: jest.fn(() =>
        Promise.resolve({
            coords: { latitude: 34.0004, longitude: -81.0359 },
        })
    ),
})); // end mock expo-location

jest.mock("react-native-maps", () => {
    const React = require("react");
    const { View } = require("react-native");
    const MockMapView = React.forwardRef((props, ref) => {
        const mockAnimateToRegion = jest.fn();
        React.useImperativeHandle(ref, () => ({
            animateToRegion: mockAnimateToRegion,
        }));

        return (
        <View
            {...props}
            ref={ref}
            testID="map-view"
            customMapStyle={props.customMapStyle}
        >
            {props.children}
        </View>
        );
    });

    return {
        __esModule: true,
        default: MockMapView,
        Marker: (props) => (
        <View
            testID="marker"
            onPress={props.onPress}
            title={props.title}
            description={props.description}
            pinColor={props.pinColor}
        >
            {props.children}
        </View>
        ),
        PROVIDER_GOOGLE: "google",
    };
}); // end mock react-native-maps

jest.mock("@expo/vector-icons/FontAwesome", () => {
    const { View, Text } = require("react-native");
    const MockIcon = (props) => (
        <View testID={props.testID || "font-awesome-icon"}>
        <Text>{props.name}</Text>
        </View>
    );
    MockIcon.isLoaded = true;
    return MockIcon;
}); // end mock FontAwesome

jest.mock("expo-font", () => ({
    useFonts: jest.fn(() => [true, null]),
})); // end mock expo-font

jest.mock("firebase/firestore", () => ({
    collection: jest.fn(),
    getDocs: jest.fn(() =>
        Promise.resolve({
        docs: [
            {
            id: "1",
            data: () => ({
                title: "1000 Catawba Street",
                catId: 9492,
                description:
                "Alternate Names: 1000 Catawba St, Columbia, SC 29201, USA, 1000, Catawba Street, Olympia-Granby, Columbia, Richland County, South Carolina, United States, 29201, 5706",
                id: 223277,
                latitude: 33.987514,
                latitudeDelta: 0.000001,
                longitude: -81.03051,
                longitudeDelta: 0.000001,
            }),
            },
            {
            id: "2",
            data: () => ({
                title: "1005 Idlewilde Boulevard",
                catId: 23396,
                description:
                "Alternate Names: 1005 Idlewilde Blvd, Columbia, SC 29201, USA, 1005, Idlewilde Boulevard, Columbia, Richland County, South Carolina, United States, 29201, 4825",
                id: 225663,
                latitude: 33.966026,
                latitudeDelta: 0.000001,
                longitude: -81.009964,
                longitudeDelta: 0.000001,
            }),
            },
            {
            id: "3",
            data: () => ({
                title: "1027 Barnwell St. (University Foundations)",
                catId: 23396,
                description:
                "Alternate Names: 1027 Barnwell St, Columbia, SC 29201, USA, 1027, Barnwell Street, University Hill, Columbia, Richland County, South Carolina, United States, 29201, 3801",
                id: 223347,
                latitude: 34.002277,
                latitudeDelta: 0.000001,
                longitude: -81.023407,
                longitudeDelta: 0.000001,
            }),
            },
        ],
        })
    ),
    getDoc: jest.fn(() =>
        Promise.resolve({ exists: () => false, data: () => ({}) })
    ),
    updateDoc: jest.fn(() => Promise.resolve()),
    addDoc: jest.fn(() => Promise.resolve({ id: "mock-id" })),
    deleteDoc: jest.fn(),
    doc: jest.fn(),
    setDoc: jest.fn(() => Promise.resolve()),
})); // end mock firestore

jest.mock("@react-native-async-storage/async-storage", () => ({
    getItem: jest.fn(),
    setItem: jest.fn(),
    removeItem: jest.fn(),
})); // end mock async-storage

jest.mock("react-native-elements", () => {
    const { View } = require("react-native");
    return {
        Icon: () => <View testID="icon" />,
    };
}); // end mock react-native-elements



//              DESCRIBE

// For Testing the Modal in HomeScreen. Namely, the Info Modal
describe("HomeScreen Modal Tests", () => {
    const theme = {
        colors: {
        text: "#000000",
        garnetWhite: "#FFFFFF",
        card: "#F5F5F5",
        },
    };

    const mockCategoryVisibility = {
        categoryVisibility: {
        9492: true,
        23396: false,
        },
        isInitialized: true,
    };
    
    const mockPush = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
        useRouter.mockReturnValue({ push: mockPush });
        useLocalSearchParams.mockReturnValue({});
    
        Location.requestForegroundPermissionsAsync.mockResolvedValue({
            status: "granted",
        });
        Location.getCurrentPositionAsync.mockResolvedValue({
            coords: { latitude: 34.0004, longitude: -81.0359 },
        });
        });
    
        // moved above. uncomment if it breaks
        /*
        const renderHomeScreen = (props = {}) => {
            return render(
                <ThemeContext.Provider value={{ theme: mockTheme }}>
                <CategoryVisibilityContext.Provider value={mockCategoryVisibility}>
                    <HomeScreen {...props} />
                </CategoryVisibilityContext.Provider>
                </ThemeContext.Provider>
            );
        };
        */

    it("renders the '+' symbol and the map-marker icon correctly", async () => {
        const { getByText, getByTestId, findByTestId } = renderHomeScreen();

        // Open the modal
        //const infoButton = getByTestId("infoButton"); //changed
        const infoButton = await findByTestId("infoButton");
        //infoButton.props.onPress();
        fireEvent.press(infoButton);

        // Check if the "+" symbol is rendered
        //expect(getByText("+")).toBeTruthy();

        const plusText = await findByTestId("plus-symbol");
        const plusStyle = Array.isArray(plusText.props.style)
            ? Object.assign({}, ...plusText.props.style.filter(Boolean))
            : plusText.props.style ?? {};

            expect(plusStyle.marginLeft).toBe(2);

        // Check if the map-marker icon is rendered
        const mapMarkerIcon = getByTestId("map-marker-icon");
        expect(mapMarkerIcon).toBeTruthy();
    });

    it("ensures the '+' symbol is aligned correctly with the map-marker icon", async () => {
        const { findByTestId, getByText, getAllByText } = renderHomeScreen();
        
        const infoButton = await findByTestId("infoButton");
        fireEvent.press(infoButton); // simulate user interaction
        
        const mapMarkerWrapper = await findByTestId("map-marker-icon");
        const plusText = await findByTestId("plus-symbol");
        
        const iconStyleRaw = mapMarkerWrapper.props?.style ?? {};
        const iconStyle = Array.isArray(iconStyleRaw)
            ? Object.assign({}, ...iconStyleRaw.filter(Boolean))
            : iconStyleRaw;
        
        const plusStyleRaw = plusText.props?.style ?? {};
        const plusStyle = Array.isArray(plusStyleRaw)
            ? Object.assign({}, ...plusStyleRaw.filter(Boolean))
            : plusStyleRaw;
        
        expect(iconStyle.marginRight).toBe(2);
        expect(plusStyle.marginLeft).toBe(2);
        
    });

    it("renders the modal content correctly", async () => {
        const { getByText, getByTestId } = renderHomeScreen();

        // Open the modal
        const infoButton = await waitFor(() => getByTestId("infoButton")); // Corrected await usage

        // Open the modal
        fireEvent.press(infoButton);

        // Check if the modal content is rendered
        expect(getByText("Button Information")).toBeTruthy();

        // had to do allat bc of the way its structured
        expect(
            getByText(/Add Custom Pin Icon - Add a custom pin to the map/)
        ).toBeTruthy();

    }); // end test "renders the modal content correctly"

    it("matches the modal snapshot", () => {
        // uncomment if something goes wrong
        //const tree = render(
        //<ThemeContext.Provider value={{ theme }}>
        //    <CategoryVisibilityContext.Provider value={mockCategoryVisibility}>
        //        <HomeScreen />
        //    </CategoryVisibilityContext.Provider>
        //</ThemeContext.Provider>
        //).toJSON();
        const tree = renderHomeScreen().toJSON();

        expect(tree).toMatchSnapshot();
    }); // end test "matches the modal snapshot"
}); // End of describe