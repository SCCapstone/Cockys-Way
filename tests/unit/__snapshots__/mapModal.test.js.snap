// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen Modal Tests matches the modal snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    color="#73000A"
    size="large"
  />
</View>
`;
